src/SectorView.cpp

```diff
@@ -47,6 +47,128 @@ enum DetailSelection {
+// cache - source of sector and system data
+// pos -  absolute coordinates of a point (in sectors, float)
+// returns the found SystemPath, (system closest to the given position)
+// if returned !SystemPath.IsSystemPath(), it means that nothing was found
+static SystemPath nearest_system_to_pos(cache_t cache, const vector3f &pos)
+{
+	// maximum size in sectors of the search box
+	const int MAX_SEARCH_DIST = 20;
+	// extract the coordinates of the sector, and the coordinates of the position within the sector
+	SystemPath sectorPos(floor(pos.x), floor(pos.y), floor(pos.z));
+	vector3f crd(FFRAC(pos.x), FFRAC(pos.y), FFRAC(pos.z));
+	crd *= Sector::SIZE; // convert to l.y.
+	// directions of "growth" of the search box, if necessary
+	box_t enlarge = { -1, 1, -1, 1, -1, 1 };
+	// remember the initial sector coordinates and point coordinates as boxes with size 0, for ease of comparison
+	box_t originBox = {
+		sectorPos.sectorX, sectorPos.sectorX,
+		sectorPos.sectorY, sectorPos.sectorY,
+		sectorPos.sectorZ, sectorPos.sectorZ
+	};
+	std::array<float, SIDES> posBox = { crd.x, crd.x, crd.y, crd.y, crd.z, crd.z };
+	// we will remember in which box we were already looked, so as not to repeat the search
+	box_t prevSearchBox = originBox;
+	// here we will accumulate the minimum distance
+	float min_dist = FLT_MAX;
+	// here we will accumulate the nearest system (this will be the result of the function)
+	SystemPath best_sys;
+	// the first iteration of the search - only in the sector where the base point is located
+	box_t searchBox = originBox;
+	++searchBox[XMAX];
+	++searchBox[YMAX];
+	++searchBox[ZMAX];
+	bool search_complete;
+	do {
+		// looking for the nearest system in the current search box
+		for (int x = searchBox[XMIN]; x < searchBox[XMAX]; ++x)
+			for (int y = searchBox[YMIN]; y < searchBox[YMAX]; ++y)
+				for (int z = searchBox[ZMIN]; z < searchBox[ZMAX]; ++z)
+					if (!in_box(x, y, z, prevSearchBox)) // we are not looking where we have already looked
+						try_sector_for_nearer_system(cache, crd, sectorPos, SystemPath(x, y, z), min_dist, best_sys);
+		// check the search results, and enlarge the search box, if necessary
+		search_complete = true;
+		prevSearchBox = searchBox;
+		if (min_dist == FLT_MAX) {
+			// still did not find the system at all - we are expanding in all directions
+			for (int side = 0; side < SIDES; ++side) {
+				searchBox[side] += enlarge[side];
+			}
+			search_complete = false;
+		} else {
+			// found the system - we check if the found distance can go beyond the search box
+			for (int side = 0; side < SIDES; ++side) {
+				float side_crd = (searchBox[side] - originBox[side]) * Sector::SIZE;
+				float dist_to_edge = pow(posBox[side] - side_crd, 2);
+				if (dist_to_edge < min_dist) {
+					// the system is closer to this edge of the search box than the nearest found - we expand in this direction
+					searchBox[side] += enlarge[side];
+					search_complete = false;
+				}
+			}
+		}
```

Comment:
I am very confused about the logic in this function; to start, `dist_to_edge` should critically mention that it is *squared* distance, as the actual LengthSqr call is squirreled away in the `try_sector_for_nearer_system` function and thus the logical assumption is that this is simply Euclidean distance, leading the `pow(..., 2)` call to misdirect the reader.

Additionally, the need (if any) to expand the *search box* to encompass a system found by *searching with that search box* is extremely counter-intuitive logic, and I find myself completely questioning its validity and existence. After a good few minutes of thinking, the only case I can derive is where the search position originates at the very extreme edge of a sector's bounds, a system is found N sectors away, and a second, nearer system is located N+1 sectors away from the originating sector. Even so, this is very non-obvious behavior not explained by the comments, and something that I would highly recommend be replaced by a search with a fixed radius.

Reply:
@Web-eWorks of course, here only the squares of distances are used, apparently I underestimated the "unusualness" of this.
About the need to expand the search box - for example, here is a search in a 3x3 box:

![artflow_202107161139](https://user-images.githubusercontent.com/18342621/125919116-c430c068-f9d0-4fa5-8dce-4bc10eee6da4.png)

Arrows indicate places where unaccounted points may be located. I think this is not a very extreme case, given that on average there can be 2-3 stars in the sector. Moreover, in three dimensions the situation is even more dramatic, because the diagonal of the cube is longer than the diagonal of the square.

Reply:
After another 20 minutes I realized that was what this code was taking into account, although to do so I had to disregard the comments entirely and rethink the entire logic flow of that sub-branch. I see what's going on there, although I'd still recommend side-stepping it with a fixed-radius search (or at least clarifying variable names and comments to explicitly mention that's the edge case you're handling).

Reply:
@Web-eWorks Please tell me which comment misled you, because it seems to me that the comments describe exactly what is happening in the code. Therefore, it is very important to see the outside view.

It seemed to me that such a problem arises when searching for the nearest in any spatial index, be it a quad-tree, or a kd-tree (or in a simple grid, as in our case), when the found nearest in a given branch may not actually be the nearest. And we need to search in neighboring branches that are closer to us than the nearest found.
About searching in a fixed sphere, I will answer elsewhere.

Reply:
Yes, the comments mention that you're expanding towards the found system rather than away; I'd appreciate the comments explicitly mentioning that this covers the edge case in which the distance to the search box edge is less than the diagonal distance to the found system.

Reply:
> apparently I underestimated the "unusualness" of this

Generally speaking, if it's not immediately apparent what distance metric you're using (e.g. a call to LengthSqr within the same function) it's better to either comment it or make it part of the variable name. Using `min_dist_sq` significantly reduces the ambiguity of the function, and overall improves the immediate readability of the code.

I'm personally trying to adopt the philosophy that the larger the scope a variable has, the longer and more descriptive its name should be. Loop indices can be a single letter, function-local variables and parameters one to two words, variables that are passed around to multiple functions more than one word, etc. If you have to read back to a variable's definition to know the semantics of the contents, you've written the equivalent of a rude note to the next guy who has to work on the code.

Aside from all that, I'm requesting one more change here - instead of `pow(v, 2)`, please prefer `v * v` - unless you're using a non-integer power argument, you can simply compose a square or cube exponentiation and it winds up faster in debug builds as it doesn't call into the fairly complicated exponentiation function.

Reply:
Done!